{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.linear_model import LinearRegression\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import seaborn as sns \n",
    "from scipy.stats import pearsonr, spearmanr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Expression via SNPs: Linear Regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_file = \"datasets/GD660.GeneQuantRPKM.txt.gz\"\n",
    "test_set_file = \"datasets/5_fold_cv_split.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in genotypes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_set = pd.read_csv(test_set_file, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>genotype</th>\n",
       "      <th>GT_encoding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>snp_6_32552087</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>snp_6_32629257</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>1|1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>snp_6_32634306</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>1|1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs9899833</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6547</th>\n",
       "      <td>snp_6_32552087</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>1|0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6548</th>\n",
       "      <td>snp_6_32629257</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>1|1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6549</th>\n",
       "      <td>snp_6_32634306</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>1|1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6550</th>\n",
       "      <td>rs9899833</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>1|1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6551</th>\n",
       "      <td>rs9912530</td>\n",
       "      <td>NA20828</td>\n",
       "      <td>0|1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6552 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID sample_id genotype  GT_encoding\n",
       "0          rs9999724   HG00096      0|0            0\n",
       "1     snp_6_32552087   HG00096      0|1            1\n",
       "2     snp_6_32629257   HG00096      1|1            2\n",
       "3     snp_6_32634306   HG00096      1|1            2\n",
       "4          rs9899833   HG00096      0|1            1\n",
       "...              ...       ...      ...          ...\n",
       "6547  snp_6_32552087   NA20828      1|0            1\n",
       "6548  snp_6_32629257   NA20828      1|1            2\n",
       "6549  snp_6_32634306   NA20828      1|1            2\n",
       "6550       rs9899833   NA20828      1|1            2\n",
       "6551       rs9912530   NA20828      0|1            1\n",
       "\n",
       "[6552 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes = pd.read_csv(\"datasets/variants_subset.tsv\", sep=\"\\t\").drop(columns=[\"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\",\t\"FORMAT\"])\n",
    "genotypes = pd.melt(genotypes, id_vars=[\"#CHROM\", \"POS\", \"ID\"], var_name=\"sample_id\", value_name=\"genotype\")\n",
    "genotypes[\"#CHROM\"] = genotypes[\"#CHROM\"].astype(str)\n",
    "genotypes[\"POS\"] = genotypes[\"POS\"].astype(str)\n",
    "genotypes.loc[genotypes.ID==\".\", 'ID'] = \"snp_\" + genotypes[\"#CHROM\"] + \"_\" + genotypes.POS\n",
    "genotypes = genotypes.drop(columns=[\"#CHROM\", \"POS\"])\n",
    "genotypes.genotype = genotypes['genotype'].str.split(':',expand=True)[0] \n",
    "conditions = [\n",
    "    (genotypes['genotype'] == '0|0'), (genotypes['genotype'] == '1|0'), (genotypes['genotype'] == '0|1'), (genotypes['genotype'] == '1|1')]\n",
    "choices = [0, 1, 1, 2]\n",
    "genotypes['GT_encoding'] = np.select(conditions, choices)\n",
    "#genotypes = pd.merge(test_set.drop(columns=[\"sample_id\", \"gene_id\"]).drop_duplicates(), genotypes, left_on=[\"patient\"], right_on=[\"sample_id\"])\n",
    "genotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in embeddings to get the GENE IDs: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "embeddings = pd.read_pickle(\"test/embeddings_complete.pickle\")\n",
    "sample_id = []\n",
    "gene_id = []\n",
    "embedding = []\n",
    "for i in range(0, len(embeddings), 2):\n",
    "    sid = embeddings[i][0].split(':')[1]\n",
    "    gid = embeddings[i][0].split(':')[2]\n",
    "    sample_id.append(sid)\n",
    "    gene_id.append(gid)\n",
    "    # combined_embedding = np.concatenate([embeddings[i][1], embeddings[i+1][1]]) # concatenate 2 embeddings\n",
    "    combined_embedding = np.array(embeddings[i][1] + embeddings[i+1][1]) # element-wise sum of 2 embeddings\n",
    "    embedding.append(combined_embedding)\n",
    "df = pd.DataFrame({\"sample_id\": sample_id, \"gene_id\": gene_id, \"embedding\": embedding})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in expression data: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_id_x</th>\n",
       "      <th>genotype</th>\n",
       "      <th>GT_encoding</th>\n",
       "      <th>sample_id_y</th>\n",
       "      <th>patient</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>test_split</th>\n",
       "      <th>expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>HG00096.1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>ENSG00000214425.1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.40868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>HG00096.1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>ENSG00000176681.9</td>\n",
       "      <td>3</td>\n",
       "      <td>6.34823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>HG00096.1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>ENSG00000179344.11</td>\n",
       "      <td>3</td>\n",
       "      <td>43.78876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>HG00096.1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>ENSG00000232629.4</td>\n",
       "      <td>3</td>\n",
       "      <td>9.96245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs9999724</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>0|0</td>\n",
       "      <td>0</td>\n",
       "      <td>HG00096.1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>ENSG00000238083.3</td>\n",
       "      <td>3</td>\n",
       "      <td>7.90419</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID sample_id_x genotype  GT_encoding sample_id_y  patient  \\\n",
       "0  rs9999724     HG00096      0|0            0   HG00096.1  HG00096   \n",
       "1  rs9999724     HG00096      0|0            0   HG00096.1  HG00096   \n",
       "2  rs9999724     HG00096      0|0            0   HG00096.1  HG00096   \n",
       "3  rs9999724     HG00096      0|0            0   HG00096.1  HG00096   \n",
       "4  rs9999724     HG00096      0|0            0   HG00096.1  HG00096   \n",
       "\n",
       "              gene_id  test_split  expression  \n",
       "0   ENSG00000214425.1           3     0.40868  \n",
       "1   ENSG00000176681.9           3     6.34823  \n",
       "2  ENSG00000179344.11           3    43.78876  \n",
       "3   ENSG00000232629.4           3     9.96245  \n",
       "4   ENSG00000238083.3           3     7.90419  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression = pd.read_csv(expression_file, sep=\"\\t\")\n",
    "expression = expression[expression[\"Gene_Symbol\"].isin(df[\"gene_id\"].values)]\n",
    "expression = pd.melt(expression, id_vars=[\"TargetID\", \"Gene_Symbol\", \"Chr\", \"Coord\"], var_name=\"sample_id\", value_name=\"expression\")\n",
    "expression[\"sample_id\"] = expression[\"sample_id\"].str.split(\".\", expand=True)[0] + \".\" + expression[\"sample_id\"].str.split(\".\", expand=True)[1]\n",
    "data = pd.merge(test_set, expression, left_on=[\"gene_id\", \"sample_id\"], right_on=[\"Gene_Symbol\", \"sample_id\"])\n",
    "data = pd.merge(genotypes, data, left_on=[\"sample_id\"], right_on=[\"patient\"])\n",
    "data = data.drop(columns=[\"Chr\", \"TargetID\", \"Coord\", \"Gene_Symbol\"])\n",
    "#data.rename(columns={\"sample_id_y\":\"sample_id\"}, inplace=True)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>PROBE_ID</th>\n",
       "      <th>CHR_SNP</th>\n",
       "      <th>CHR_GENE</th>\n",
       "      <th>SNPpos</th>\n",
       "      <th>TSSpos</th>\n",
       "      <th>distance</th>\n",
       "      <th>rvalue</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>log10pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55450</th>\n",
       "      <td>rs10020034</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>120294331.0</td>\n",
       "      <td>120299287</td>\n",
       "      <td>4956.0</td>\n",
       "      <td>0.548974</td>\n",
       "      <td>1.079312e-30</td>\n",
       "      <td>29.966853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55451</th>\n",
       "      <td>rs3955380</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>120334410.0</td>\n",
       "      <td>120299287</td>\n",
       "      <td>35123.0</td>\n",
       "      <td>0.547157</td>\n",
       "      <td>1.832933e-30</td>\n",
       "      <td>29.736854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55452</th>\n",
       "      <td>rs28403131</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>120285806.0</td>\n",
       "      <td>120299287</td>\n",
       "      <td>13481.0</td>\n",
       "      <td>0.545653</td>\n",
       "      <td>2.834881e-30</td>\n",
       "      <td>29.547465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55453</th>\n",
       "      <td>rs79262354</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>120338335.0</td>\n",
       "      <td>120299287</td>\n",
       "      <td>39048.0</td>\n",
       "      <td>0.545126</td>\n",
       "      <td>3.300820e-30</td>\n",
       "      <td>29.481378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55454</th>\n",
       "      <td>rs78958430</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>ENSG00000229450.2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>120338339.0</td>\n",
       "      <td>120299287</td>\n",
       "      <td>39052.0</td>\n",
       "      <td>0.544934</td>\n",
       "      <td>3.488677e-30</td>\n",
       "      <td>29.457339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP_ID ID            GENE_ID           PROBE_ID  CHR_SNP  CHR_GENE  \\\n",
       "55450  rs10020034  -  ENSG00000229450.2  ENSG00000229450.2        4         4   \n",
       "55451   rs3955380  -  ENSG00000229450.2  ENSG00000229450.2        4         4   \n",
       "55452  rs28403131  -  ENSG00000229450.2  ENSG00000229450.2        4         4   \n",
       "55453  rs79262354  -  ENSG00000229450.2  ENSG00000229450.2        4         4   \n",
       "55454  rs78958430  -  ENSG00000229450.2  ENSG00000229450.2        4         4   \n",
       "\n",
       "            SNPpos     TSSpos  distance    rvalue        pvalue  log10pvalue  \n",
       "55450  120294331.0  120299287    4956.0  0.548974  1.079312e-30    29.966853  \n",
       "55451  120334410.0  120299287   35123.0  0.547157  1.832933e-30    29.736854  \n",
       "55452  120285806.0  120299287   13481.0  0.545653  2.834881e-30    29.547465  \n",
       "55453  120338335.0  120299287   39048.0  0.545126  3.300820e-30    29.481378  \n",
       "55454  120338339.0  120299287   39052.0  0.544934  3.488677e-30    29.457339  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reference = pd.read_csv(\"datasets/EUR373.gene.cis.FDR5.all.rs137.txt.gz\", sep=\"\\t\")\n",
    "reference = reference[reference.GENE_ID.isin(data.gene_id.unique())]\n",
    "reference.GENE_ID = reference.GENE_ID.astype('category')\n",
    "reference.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_snps = reference[[\"GENE_ID\", \"SNP_ID\", \"rvalue\"]].groupby(\"GENE_ID\").max()\n",
    "best_snps = best_snps.reset_index()\n",
    "best_snps = best_snps[[\"GENE_ID\", \"SNP_ID\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only keep most predictive SNP for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.merge(data, best_snps, left_on=[\"ID\", \"gene_id\"], right_on=[\"SNP_ID\", \"GENE_ID\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "splits_train = {}\n",
    "splits_test = {}\n",
    "for i in range(1, 6):\n",
    "    r_values_train = {}\n",
    "    r_values_test = {}\n",
    "    for gene in data.gene_id.unique():\n",
    "        # extract data\n",
    "        gene_df = data[data.gene_id==gene]\n",
    "        gene_df_train = gene_df[gene_df.test_split != i]\n",
    "        gene_df_test = gene_df[gene_df.test_split == i]\n",
    "\n",
    "        X_train = np.array(gene_df_train[\"GT_encoding\"]).reshape(-1, 1)\n",
    "        X_test = np.array(gene_df_test[\"GT_encoding\"]).reshape(-1, 1)\n",
    "        y_train = np.array(gene_df_train.expression)\n",
    "        y_test = np.array(gene_df_test.expression)\n",
    "\n",
    "        # train model\n",
    "        model = LinearRegression().fit(X_train, y_train)\n",
    "\n",
    "        # evaluate model on test set\n",
    "        y_pred_train = model.predict(X_train)\n",
    "        y_pred_test = model.predict(X_test)\n",
    "\n",
    "        r_values_train[gene] = pearsonr(y_train, y_pred_train)[0]\n",
    "        r_values_test[gene] = pearsonr(y_test, y_pred_test)[0]\n",
    "    splits_train[i] = r_values_train\n",
    "    splits_test[i] = r_values_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'ENSG00000229450.2': 0.5838158873780467,\n",
       "  'ENSG00000196126.6': -0.010763378614997422,\n",
       "  'ENSG00000198502.5': 0.10200379539571876,\n",
       "  'ENSG00000179344.11': 0.4347964807105317,\n",
       "  'ENSG00000232629.4': 0.5652492282985153,\n",
       "  'ENSG00000237541.3': 0.6617140720179402,\n",
       "  'ENSG00000196735.6': 0.3548608619011706,\n",
       "  'ENSG00000214425.1': 0.3825433425358211,\n",
       "  'ENSG00000176681.9': 0.2777803066581089,\n",
       "  'ENSG00000238083.3': 0.16879455130923565},\n",
       " 2: {'ENSG00000229450.2': 0.4691115855795315,\n",
       "  'ENSG00000196126.6': 0.27985575993872325,\n",
       "  'ENSG00000198502.5': 0.07079277051916556,\n",
       "  'ENSG00000179344.11': 0.5890411023450833,\n",
       "  'ENSG00000232629.4': 0.4018050698611545,\n",
       "  'ENSG00000237541.3': 0.44436281545505874,\n",
       "  'ENSG00000196735.6': 0.560246279084256,\n",
       "  'ENSG00000214425.1': 0.32156282683694326,\n",
       "  'ENSG00000176681.9': 0.07276634100342154,\n",
       "  'ENSG00000238083.3': 0.038449204012593045},\n",
       " 3: {'ENSG00000229450.2': 0.4413664714953019,\n",
       "  'ENSG00000196126.6': 0.375531038379767,\n",
       "  'ENSG00000198502.5': -0.06995265795704717,\n",
       "  'ENSG00000179344.11': 0.37119436677140594,\n",
       "  'ENSG00000232629.4': 0.3539578404376757,\n",
       "  'ENSG00000237541.3': 0.5469369162280131,\n",
       "  'ENSG00000196735.6': 0.23847802336969912,\n",
       "  'ENSG00000214425.1': 0.5411188021979161,\n",
       "  'ENSG00000176681.9': 0.254630072547255,\n",
       "  'ENSG00000238083.3': 0.14723575713950418},\n",
       " 4: {'ENSG00000229450.2': 0.5017599867426142,\n",
       "  'ENSG00000196126.6': 0.03444543746476007,\n",
       "  'ENSG00000198502.5': 0.0877012316067547,\n",
       "  'ENSG00000179344.11': 0.5138567765204525,\n",
       "  'ENSG00000232629.4': 0.5890701768419196,\n",
       "  'ENSG00000237541.3': 0.5398267549051448,\n",
       "  'ENSG00000196735.6': 0.513807232343771,\n",
       "  'ENSG00000214425.1': 0.259837785921484,\n",
       "  'ENSG00000176681.9': 0.5361279950360627,\n",
       "  'ENSG00000238083.3': 0.49233885069583444},\n",
       " 5: {'ENSG00000229450.2': 0.5102082899938314,\n",
       "  'ENSG00000196126.6': 0.30359976443086106,\n",
       "  'ENSG00000198502.5': 0.21555297676440266,\n",
       "  'ENSG00000179344.11': 0.5388582266219795,\n",
       "  'ENSG00000232629.4': 0.1748813993133373,\n",
       "  'ENSG00000237541.3': 0.5286988082663079,\n",
       "  'ENSG00000196735.6': 0.5598978389154952,\n",
       "  'ENSG00000214425.1': 0.2608355481723357,\n",
       "  'ENSG00000176681.9': 0.32056742412190664,\n",
       "  'ENSG00000238083.3': 0.2587472620359317}}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splits_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract and plot data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = []\n",
    "g = []\n",
    "r = []\n",
    "for k, res in splits_test.items():\n",
    "    s += [k]*len(res)\n",
    "    g += list(res.keys())\n",
    "    r += list(res.values())\n",
    "test_df = pd.DataFrame({\"split\": s, \"gene_id\": g, \"rvalue\": r})\n",
    "\n",
    "s = []\n",
    "g = []\n",
    "r = []\n",
    "for k, res in splits_train.items():\n",
    "    s += [k]*len(res)\n",
    "    g += list(res.keys())\n",
    "    r += list(res.values())\n",
    "train_df = pd.DataFrame({\"split\": s, \"gene_id\": g, \"rvalue\": r})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "x1 = test_df.groupby(\"gene_id\").mean()\n",
    "x1 = x1.reset_index()\n",
    "x2 = test_df.groupby(\"gene_id\").std()\n",
    "x2 = x2.reset_index()\n",
    "x = pd.merge(x1[[\"gene_id\", \"rvalue\"]], x2[[\"gene_id\", \"rvalue\"]], on=\"gene_id\")\n",
    "x[\"method\"] = \"SNP\"\n",
    "x.to_csv(\"snp.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>split</th>\n",
       "      <th>rvalue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000176681.9</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.165889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000179344.11</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.086497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000196126.6</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.172982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000196735.6</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.143172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198502.5</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.101835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000214425.1</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.116644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229450.2</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.053658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000232629.4</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.169125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237541.3</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.077530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238083.3</th>\n",
       "      <td>1.581139</td>\n",
       "      <td>0.170690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       split    rvalue\n",
       "gene_id                               \n",
       "ENSG00000176681.9   1.581139  0.165889\n",
       "ENSG00000179344.11  1.581139  0.086497\n",
       "ENSG00000196126.6   1.581139  0.172982\n",
       "ENSG00000196735.6   1.581139  0.143172\n",
       "ENSG00000198502.5   1.581139  0.101835\n",
       "ENSG00000214425.1   1.581139  0.116644\n",
       "ENSG00000229450.2   1.581139  0.053658\n",
       "ENSG00000232629.4   1.581139  0.169125\n",
       "ENSG00000237541.3   1.581139  0.077530\n",
       "ENSG00000238083.3   1.581139  0.170690"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.groupby(\"gene_id\").std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mean: 0.35419810599101403\n",
      "SD: 0.15248618932782093\n"
     ]
    }
   ],
   "source": [
    "print(f\"mean: {test_df.groupby('gene_id').mean().rvalue.mean()}\")\n",
    "print(f\"SD: {test_df.groupby('gene_id').mean().rvalue.std()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df[\"method\"] = \"SNP\"\n",
    "test_df.to_csv(\"snp.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gaushvi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
